### ISSUE TITLE
Advice on constructing training/test dataset with labels synthesized from two files. 

### ISSUE THREAD

**Issue by mbkennel** [deeplearning4j/DataVec#155]
Gentlemen, I have the following scenario: 

I have two datasets, assume space-separated ASCII floating point, of features, stored as ordinary files on disc. One dataset is class '0', the other dataset is class '1'.   Note: there is NO column denoting the class, the class is implicitly represented by data being in one file vs the other. 

Question:  how do I appropriately read this data and construct a training dataset to train a binary classifier with one output?  (the problem clearly generalizes to > 2 classes, but my particular problem is binary classification). 

Question 2:  How do ensure that before training, either physically or implicitly, the examples are shuffled so that I don't get minibatches of all class 0 and all class 1's, as if the two datasets were naively read in and stacked in a single feature matrix?

**Comment by agibsonccc** [comment]
So the first question I have: Is this a binary file format?   We could build a record reader that "understood" this format if I knew more about it. So let's expand on that first.

1: We can build a record reader for this. The number of labels and the like don't really matter.
2: This is handled elsewhere. You would either build separate iterators or use: https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/dataset/BalanceMinibatchesTest.java#L22

**Comment by mbkennel** [comment]
1) No, the file format is not binary in my current use case, but it could be later. I'm just at the start of R&D, for now assume CSV or whitespace separated vectors, simple. 

At the moment, I see the problem not as needing a new reader from raw data, but a way to combine the results of readers on multiple files/sources, and append a target (my use case) or a feature (some other use cases, not mine) corresponding to the identity of the input file/source.

2)  This seems unnecessarily difficult for a very common use case. Let me clarify:  I'm looking to randomize ordering of features, compared to features as read in from data.  In many practical cases, the training data itself has systematic effects from early examples to late, for instance if it comes out of transactions ordered by customer number as extracted from a DB ETL, and low-numbered customers are different from high-numbered customers in systematic ways (like being older, or coming from different bank portfolios---real life cases for me)

My current situation is another one, where examples of all one class come before members of the next class because they were read in from separate files.  In both cases, if we don't do anything, then minibatches of sequential blocks of examples will not have randomly distributed examples, and training will be poor as there is excessive correlation between examples in a minibatch.   I assume, that for cache locality and to be able to use matrix libraries directly, the minibatches during training are taken from contiguous blocks of examples.  (If this is not the case, that there is a random sampling from the entire dataset and extraction into a minibatch in a temp buffer for each training eval & update, then this concern doesn't apply). 

I'm not (yet) asking for stratified sampling, where minibatches are constructed with nearly equal class proportions from separately randomized orderings of each class example.  (Is your notation of "balanced" the same as "stratified sampling")? http://stattrek.com/statistics/dictionary.aspx?definition=Stratified%20sampling

What I'm asking for is a step in which after reading in data, the examples are shuffled according to a uniformly drawn random permutation of 1...N for N being the number of train examples.  (And test examples also shuffled too, though this is usually less important).  Now, any sequential block of examples is composed of examples which occurred in random orderings of the original dataset. 

Background:  I'm the author of my company's internal nnet trainer back in 2008-2009.  It is rather hardcoded to one legacy product's workflow, but we're looking to more general problems now and there are packages available now that weren't then.  

In that trainer, I read in the dataset, and applied a 'full shuffle' of examples (which is expensive as it involves randomly located swaps of vectors across full memory).  This involves generating a random permutation of integers 0...N-1 by the standard methods, and then applying this permutation to shuffle the full dataset.  The goal is not to need a duplicate of the dataset (as can be close to max  RAM available).  It is possible to apply a permutation to a dataset requiring just one vector of temp storage---it is a student problem in one of the Knuth books (!) and I've dug up the algorithm somewhere and can share it if isn't well known to people now.  

This order shuffling is what I'm asking for in DL4J/datavec/nd4j as it would be unnecessarily expensive and undesirable to have to do it externally when data are large and eventually less and less preprocessed.

Further training question:  how do you order minibatches? 

In my code: During each training epoch there are B=floor(N/N_B) minibatches of size N_B examples plus potentially a smaller runt minibatch. I then make a random permutation of 0..B-1 and do updates on minibatches in that ordering.  So every epoch chooses a random order of minibatches (which is cheap---change starting location), but not a full shuffle of individual data which is expensive.  Multithreading would complicate this of course but it's still a matter of choosing minibatches from a randomized pool. 

Every 10 epochs I do a full reshuffle of all data, as was done upon read-in.  This is about as costly in time as one training epoch.
